python/ql/test/experimental/dataflow/typetracking/moduleattr.ql

```diff
@@ -2,22 +2,24 @@ import python
 import semmle.python.dataflow.new.DataFlow
 import semmle.python.dataflow.new.TypeTracker
 
-DataFlow::Node module_tracker(TypeTracker t) {
+DataFlow::LocalSourceNode module_tracker(TypeTracker t) {
   t.start() and
   result = DataFlow::importNode("module")
   or
   exists(TypeTracker t2 | result = module_tracker(t2).track(t2, t))
 }
 
-query DataFlow::Node module_tracker() { result = module_tracker(DataFlow::TypeTracker::end()) }
+query DataFlow::Node module_tracker() {
+  module_tracker(DataFlow::TypeTracker::end()).flowsTo(result)
```

Comment:
Why do we need this `.flowsTo(result)`?

I'm slightly confused about it, since JS doesn't do this in the [firebase type-trackers](https://github.com/github/codeql/blob/408ac2729d2330293f6fad6af922083159bfc426/javascript/ql/src/semmle/javascript/frameworks/Firebase.qll#L22-L23) (honestly didn't look at any other ones). -- is our situation so different now that our base template for doing type-tracking ends up being different?

Reply:
The reason JS gets away with not doing this is that they have a rich and varied selection of utility methods on `SourceNode` that internally take care of the local flow details. That's why most of the JS type trackers work in terms of `SourceNode`s exclusively.

In this case, however, I opted for `flowsTo` simply because otherwise the test output would change quite a bit more. If we only type track to `LocalSourceNode`s, we'll only get the first occurrence of a tracked node, but our tests have annotations for _all_ occurrences.

Reply:
Okay. That makes sense, thanks for a nice explanation.

I was considering whether a long term goal would be that results of both public/private type-tracking predicates are `DataFlow::LocalSourceNode`, but I think that would require us to have the same _rich and varied selection of utility methods_, so we don't end up in a situation where it's very easy to forget using `.flowsTo`. So for now, I guess I will adjust my template for how to write type-trackers in Python :+1: